File: lib/sidereon/clock_stability.ex
defmodule Sidereon.ClockStability do
  @moduledoc """
  Allan-family clock-stability estimators.

  The functions in this module delegate to `sidereon-core`'s IEEE-1139 Allan,
  modified Allan, Hadamard, and time-deviation estimators. Phase deviations are
  in seconds, fractional-frequency samples are dimensionless, and averaging
  times are seconds.
  """

  alias __MODULE__.Curve
  alias __MODULE__.Curves
  alias Sidereon.GNSS.RINEX.Observations
  alias Sidereon.NIF

  defmodule Curve do
    @moduledoc """
    One clock-stability estimator curve.

    `:tau_s` is the averaging time in seconds, `:deviation` is the estimator
    value, and `:n` is the number of estimator terms used at each point.
    """

    @enforce_keys [:tau_s, :deviation, :n]
    defstruct [:tau_s, :deviation, :n]

    @type t :: %__MODULE__{
            tau_s: [float()],
            deviation: [float()],
            n: [non_neg_integer()]
          }
  end

  defmodule Curves do
    @moduledoc """
    Combined Allan-family output.

    Each field is either a `Sidereon.ClockStability.Curve` or `nil` when that
    estimator was not requested. Deviations are dimensionless except `:tdev`,
    whose deviation values are seconds.
    """

    alias Sidereon.ClockStability.Curve

    defstruct [:adev, :overlapping_adev, :mdev, :hdev, :tdev]

    @type t :: %__MODULE__{
            adev: Curve.t() | nil,
            overlapping_adev: Curve.t() | nil,
            mdev: Curve.t() | nil,
            hdev: Curve.t() | nil,
            tdev: Curve.t() | nil
          }
  end

  @type series ::
          [number()]
          | {:phase_seconds, [number()]}
          | {:fractional_frequency, [number()]}
          | {:phase_seconds_with_gaps, [number() | nil]}
          | {:fractional_frequency_with_gaps, [number() | nil]}

  @type tau_grid :: :octave | :all | {:explicit, [pos_integer()]}
  @type gap_policy :: :reject | :omit_terms
  @type estimator :: :adev | :overlapping_adev | :mdev | :hdev | :tdev
  @type allan_error ::
          :empty_series
          | :invalid_tau0
          | :no_estimators
          | :empty_tau_grid
          | :invalid_averaging_factor
          | :too_few_samples
          | :non_finite_sample
          | :gap
          | :non_finite_tau
          | :non_finite_deviation
          | term()

  @doc """
  Plain non-overlapping Allan deviation for explicit averaging factors.

  `tau0_s` is the base sample interval in seconds. `averaging_factors` are the
  integer `m` values, so each output `tau_s` is `m * tau0_s`.
  """
  @spec allan_deviation(series(), number(), [pos_integer()]) :: {:ok, Curve.t()} | {:error, allan_error()}
  def allan_deviation(series, tau0_s, averaging_factors) do
    estimator(:adev, series, tau0_s, averaging_factors)
  end

  @doc """
  Fully overlapping Allan deviation for explicit averaging factors.

  `tau0_s` and returned `tau_s` values are seconds.
  """
  @spec overlapping_adev(series(), number(), [pos_integer()]) :: {:ok, Curve.t()} | {:error, allan_error()}
  def overlapping_adev(series, tau0_s, averaging_factors) do
    estimator(:overlapping_adev, series, tau0_s, averaging_factors)
  end

  @doc """
  Modified Allan deviation for explicit averaging factors.

  `tau0_s` and returned `tau_s` values are seconds.
  """
  @spec modified_adev(series(), number(), [pos_integer()]) :: {:ok, Curve.t()} | {:error, allan_error()}
  def modified_adev(series, tau0_s, averaging_factors) do
    estimator(:mdev, series, tau0_s, averaging_factors)
  end

  @doc """
  Overlapping Hadamard deviation for explicit averaging factors.

  `tau0_s` and returned `tau_s` values are seconds.
  """
  @spec hadamard_deviation(series(), number(), [pos_integer()]) :: {:ok, Curve.t()} | {:error, allan_error()}
  def hadamard_deviation(series, tau0_s, averaging_factors) do
    estimator(:hdev, series, tau0_s, averaging_factors)
  end

  @doc """
  Time deviation for explicit averaging factors.

  Returned deviation values are seconds.
  """
  @spec time_deviation(series(), number(), [pos_integer()]) :: {:ok, Curve.t()} | {:error, allan_error()}
  def time_deviation(series, tau0_s, averaging_factors) do
    estimator(:tdev, series, tau0_s, averaging_factors)
  end

  @doc """
  Compute a configured set of Allan-family curves.

  Options:

    * `:estimators` - `:standard` for OADEV, MDEV, HDEV, and TDEV, `:all`,
      `:none`, a list of estimator atoms, or a map of estimator booleans.
    * `:tau_grid` - `:octave`, `:all`, or `{:explicit, factors}`. Averaging
      times are seconds.
    * `:gap_policy` - `:reject` or `:omit_terms` for gapped series.
  """
  @spec compute_allan_deviations(series(), number(), keyword()) :: {:ok, Curves.t()} | {:error, allan_error()}
  def compute_allan_deviations(series, tau0_s, opts \\ []) do
    with {:ok, kind, samples} <- normalize_series(series),
         {:ok, estimator_flags} <- estimator_flags(Keyword.get(opts, :estimators, :standard)),
         {:ok, tau_kind, tau_factors} <- tau_grid(Keyword.get(opts, :tau_grid, :octave)),
         {:ok, gap_policy} <- gap_policy(Keyword.get(opts, :gap_policy, :reject)) do
      case NIF.clock_compute_allan_deviations(
             kind,
             samples,
             tau0_s / 1.0,
             estimator_flags,
             tau_kind,
             tau_factors,
             gap_policy
           ) do
        {:ok, curves} -> {:ok, curves(curves)}
        {:error, _} = err -> err
        other -> {:error, other}
      end
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Extract RINEX receiver-clock phase deviations in seconds.

  Event epochs are returned as `nil` gaps. The result can be passed as
  `{:phase_seconds_with_gaps, values}` to `compute_allan_deviations/3`.
  """
  @spec receiver_clock_phase_deviations(Observations.t()) :: [float() | nil]
  def receiver_clock_phase_deviations(%Observations{handle: handle}) do
    NIF.clock_receiver_phase_deviations(handle)
  end

  defp estimator(kind, series, tau0_s, averaging_factors) do
    with {:ok, series_kind, samples} <- normalize_series(series),
         {:ok, factors} <- averaging_factors(averaging_factors) do
      case NIF.clock_allan_estimator(Atom.to_string(kind), series_kind, samples, tau0_s / 1.0, factors) do
        {:ok, curve} -> {:ok, curve(curve)}
        {:error, _} = err -> err
        other -> {:error, other}
      end
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  defp normalize_series(values) when is_list(values), do: normalize_series({:phase_seconds, values})

  defp normalize_series({kind, values}) when kind in [:phase_seconds, :fractional_frequency] and is_list(values) do
    values
    |> Enum.reduce_while({:ok, []}, fn
      value, {:ok, acc} when is_number(value) -> {:cont, {:ok, [value / 1.0 | acc]}}
      _value, _acc -> {:halt, {:error, :invalid_sample}}
    end)
    |> reverse_samples(kind)
  end

  defp normalize_series({kind, values})
       when kind in [:phase_seconds_with_gaps, :fractional_frequency_with_gaps] and is_list(values) do
    values
    |> Enum.reduce_while({:ok, []}, fn
      nil, {:ok, acc} -> {:cont, {:ok, [nil | acc]}}
      value, {:ok, acc} when is_number(value) -> {:cont, {:ok, [value / 1.0 | acc]}}
      _value, _acc -> {:halt, {:error, :invalid_sample}}
    end)
    |> reverse_samples(kind)
  end

  defp normalize_series(_other), do: {:error, :invalid_series}

  defp reverse_samples({:ok, samples}, kind), do: {:ok, Atom.to_string(kind), Enum.reverse(samples)}
  defp reverse_samples({:error, _} = err, _kind), do: err

  defp averaging_factors(values) when is_list(values) do
    values
    |> Enum.reduce_while({:ok, []}, fn
      value, {:ok, acc} when is_integer(value) and value > 0 -> {:cont, {:ok, [value | acc]}}
      _value, _acc -> {:halt, {:error, :invalid_averaging_factor}}
    end)
    |> case do
      {:ok, factors} -> {:ok, Enum.reverse(factors)}
      {:error, _} = err -> err
    end
  end

  defp averaging_factors(_other), do: {:error, :invalid_averaging_factor}

  defp estimator_flags(:standard), do: {:ok, {false, true, true, true, true}}
  defp estimator_flags(:all), do: {:ok, {true, true, true, true, true}}
  defp estimator_flags(:none), do: {:ok, {false, false, false, false, false}}

  defp estimator_flags(values) when is_list(values) do
    valid = [:adev, :overlapping_adev, :mdev, :hdev, :tdev]

    if Enum.all?(values, &(&1 in valid)) do
      {:ok,
       {
         :adev in values,
         :overlapping_adev in values,
         :mdev in values,
         :hdev in values,
         :tdev in values
       }}
    else
      {:error, :invalid_estimators}
    end
  end

  defp estimator_flags(values) when is_map(values) do
    {:ok,
     {
       Map.get(values, :adev, false),
       Map.get(values, :overlapping_adev, false),
       Map.get(values, :mdev, false),
       Map.get(values, :hdev, false),
       Map.get(values, :tdev, false)
     }}
  end

  defp estimator_flags(_other), do: {:error, :invalid_estimators}

  defp tau_grid(:octave), do: {:ok, "octave", []}
  defp tau_grid(:all), do: {:ok, "all", []}

  defp tau_grid({:explicit, factors}) do
    with {:ok, factors} <- averaging_factors(factors) do
      {:ok, "explicit", factors}
    end
  end

  defp tau_grid(_other), do: {:error, :invalid_tau_grid}

  defp gap_policy(:reject), do: {:ok, "reject"}
  defp gap_policy(:omit_terms), do: {:ok, "omit_terms"}
  defp gap_policy(_other), do: {:error, :invalid_gap_policy}

  defp curve(%{tau_s: tau_s, deviation: deviation, n: n}) do
    %Curve{tau_s: tau_s, deviation: deviation, n: n}
  end

  defp curves(raw) do
    %Curves{
      adev: maybe_curve(raw.adev),
      overlapping_adev: maybe_curve(raw.overlapping_adev),
      mdev: maybe_curve(raw.mdev),
      hdev: maybe_curve(raw.hdev),
      tdev: maybe_curve(raw.tdev)
    }
  end

  defp maybe_curve(nil), do: nil
  defp maybe_curve(raw), do: curve(raw)
end

File: native/sidereon_nif/src/clock_stability.rs
//! Rustler boundary for Allan-family clock-stability estimators.
//!
//! The estimator numerics live in `sidereon_core::clock_stability`. This module
//! decodes tagged sample series and option terms, calls the core functions, and
//! encodes the result curves. Numeric samples cross as ordinary list terms.

use crate::rinex_obs::RinexObsResource;
use rustler::{Encoder, Env, Error, NifResult, ResourceArc, Term};
use sidereon_core::clock_stability::{
    allan_deviation, allan_deviation_power_law_slope, compute_allan_deviations,
    fit_power_law_noise, hadamard_deviation, modified_adev,
    modified_allan_deviation_power_law_slope, overlapping_adev, receiver_clock_phase_deviations,
    time_deviation, AllanDeviationCurves, AllanError, AllanEstimator, AllanEstimatorSet,
    AllanInput, AllanOptions, AllanResult, AllanSeries, GapPolicy, PowerLawNoiseError,
    PowerLawNoiseFit, PowerLawNoiseOptions, PowerLawNoiseRegion, PowerLawNoiseType, PowerLawOctave,
    PowerLawOctaveDominance, PowerLawOctaveFlag, TauGrid,
};

mod atoms {
    rustler::atoms! {
        ok,
        error,
        empty_series,
        invalid_tau0,
        no_estimators,
        empty_tau_grid,
        invalid_averaging_factor,
        too_few_samples,
        non_finite_sample,
        gap,
        non_finite_tau,
        non_finite_deviation,
        invalid_options,
        invalid_curve
    }
}

#[derive(Debug, Clone, rustler::NifMap)]
struct AllanResultTerm {
    tau_s: Vec<f64>,
    deviation: Vec<f64>,
    n: Vec<i64>,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct AllanCurvesTerm {
    adev: Option<AllanResultTerm>,
    overlapping_adev: Option<AllanResultTerm>,
    mdev: Option<AllanResultTerm>,
    hdev: Option<AllanResultTerm>,
    tdev: Option<AllanResultTerm>,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct PowerLawOctaveTerm {
    tau_start_s: f64,
    tau_end_s: f64,
    point_count: i64,
    adev_slope: Option<f64>,
    mdev_slope: Option<f64>,
    slope_scatter: Option<f64>,
    dominance: String,
    noise_type: Option<String>,
    flag: Option<String>,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct PowerLawRegionTerm {
    noise_type: String,
    tau_start_s: f64,
    tau_end_s: f64,
    octave_count: i64,
    point_count: i64,
    mean_slope: f64,
    coefficient: f64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct PowerLawFitTerm {
    dominant_per_octave: Vec<PowerLawOctaveTerm>,
    coefficients: Vec<f64>,
    regions: Vec<PowerLawRegionTerm>,
}

type PowerLawOptionsTerm = (usize, f64, f64, f64, f64);

enum SeriesOwned {
    Values(Vec<f64>),
    Gapped(Vec<Option<f64>>),
}

impl SeriesOwned {
    fn as_series(&self, kind: &str) -> NifResult<AllanSeries<'_>> {
        match (kind, self) {
            ("phase_seconds", Self::Values(values)) => Ok(AllanSeries::PhaseSeconds(values)),
            ("fractional_frequency", Self::Values(values)) => {
                Ok(AllanSeries::FractionalFrequency(values))
            }
            ("phase_seconds_with_gaps", Self::Gapped(values)) => {
                Ok(AllanSeries::PhaseSecondsWithGaps(values))
            }
            ("fractional_frequency_with_gaps", Self::Gapped(values)) => {
                Ok(AllanSeries::FractionalFrequencyWithGaps(values))
            }
            _ => Err(Error::Term(Box::new("unknown Allan series kind"))),
        }
    }
}

fn series_owned(kind: &str, samples: Vec<Option<f64>>) -> NifResult<SeriesOwned> {
    match kind {
        "phase_seconds" | "fractional_frequency" => {
            let mut values = Vec::with_capacity(samples.len());
            for value in samples {
                values.push(value.ok_or_else(|| Error::Term(Box::new("unexpected gap")))?);
            }
            Ok(SeriesOwned::Values(values))
        }
        "phase_seconds_with_gaps" | "fractional_frequency_with_gaps" => {
            Ok(SeriesOwned::Gapped(samples))
        }
        _ => Err(Error::Term(Box::new("unknown Allan series kind"))),
    }
}

fn tau_grid(kind: &str, factors: Vec<usize>) -> NifResult<TauGrid> {
    match kind {
        "octave" => Ok(TauGrid::Octave),
        "all" => Ok(TauGrid::All),
        "explicit" => Ok(TauGrid::Explicit(factors)),
        _ => Err(Error::Term(Box::new("unknown Allan tau grid"))),
    }
}

fn gap_policy(kind: &str) -> NifResult<GapPolicy> {
    match kind {
        "reject" => Ok(GapPolicy::Reject),
        "omit_terms" => Ok(GapPolicy::OmitTerms),
        _ => Err(Error::Term(Box::new("unknown Allan gap policy"))),
    }
}

fn estimator_set(
    (adev, overlapping_adev, mdev, hdev, tdev): (bool, bool, bool, bool, bool),
) -> AllanEstimatorSet {
    AllanEstimatorSet {
        adev,
        overlapping_adev,
        mdev,
        hdev,
        tdev,
    }
}

fn result_term(result: AllanResult) -> AllanResultTerm {
    AllanResultTerm {
        tau_s: result.tau_s,
        deviation: result.deviation,
        n: result.n.into_iter().map(|value| value as i64).collect(),
    }
}

fn result_from_term(term: AllanResultTerm) -> NifResult<AllanResult> {
    if term.n.iter().any(|value| *value < 0) {
        return Err(Error::Term(Box::new("negative Allan term count")));
    }
    Ok(AllanResult {
        tau_s: term.tau_s,
        deviation: term.deviation,
        n: term.n.into_iter().map(|value| value as usize).collect(),
    })
}

fn curves_term(curves: AllanDeviationCurves) -> AllanCurvesTerm {
    AllanCurvesTerm {
        adev: curves.adev.map(result_term),
        overlapping_adev: curves.overlapping_adev.map(result_term),
        mdev: curves.mdev.map(result_term),
        hdev: curves.hdev.map(result_term),
        tdev: curves.tdev.map(result_term),
    }
}

fn error_atom(error: AllanError) -> rustler::Atom {
    match error {
        AllanError::EmptySeries => atoms::empty_series(),
        AllanError::InvalidTau0 { .. } => atoms::invalid_tau0(),
        AllanError::NoEstimators => atoms::no_estimators(),
        AllanError::EmptyTauGrid => atoms::empty_tau_grid(),
        AllanError::InvalidAveragingFactor { .. } => atoms::invalid_averaging_factor(),
        AllanError::TooFewSamples { .. } => atoms::too_few_samples(),
        AllanError::NonFiniteSample { .. } => atoms::non_finite_sample(),
        AllanError::Gap { .. } => atoms::gap(),
        AllanError::NonFiniteTau { .. } => atoms::non_finite_tau(),
        AllanError::NonFiniteDeviation { .. } => atoms::non_finite_deviation(),
    }
}

fn estimator(kind: &str) -> NifResult<AllanEstimator> {
    match kind {
        "adev" => Ok(AllanEstimator::Adev),
        "overlapping_adev" => Ok(AllanEstimator::OverlappingAdev),
        "mdev" => Ok(AllanEstimator::Mdev),
        "hdev" => Ok(AllanEstimator::Hdev),
        "tdev" => Ok(AllanEstimator::Tdev),
        _ => Err(Error::Term(Box::new("unknown Allan estimator"))),
    }
}

fn noise_type(kind: &str) -> NifResult<PowerLawNoiseType> {
    match kind {
        "random_walk_fm" => Ok(PowerLawNoiseType::RandomWalkFM),
        "flicker_fm" => Ok(PowerLawNoiseType::FlickerFM),
        "white_fm" => Ok(PowerLawNoiseType::WhiteFM),
        "flicker_pm" => Ok(PowerLawNoiseType::FlickerPM),
        "white_pm" => Ok(PowerLawNoiseType::WhitePM),
        _ => Err(Error::Term(Box::new("unknown power-law noise type"))),
    }
}

fn noise_type_string(value: PowerLawNoiseType) -> String {
    match value {
        PowerLawNoiseType::RandomWalkFM => "random_walk_fm",
        PowerLawNoiseType::FlickerFM => "flicker_fm",
        PowerLawNoiseType::WhiteFM => "white_fm",
        PowerLawNoiseType::FlickerPM => "flicker_pm",
        PowerLawNoiseType::WhitePM => "white_pm",
    }
    .to_string()
}

fn flag_string(value: PowerLawOctaveFlag) -> String {
    match value {
        PowerLawOctaveFlag::UnderSampled => "under_sampled",
        PowerLawOctaveFlag::DegenerateDeviation => "degenerate_deviation",
        PowerLawOctaveFlag::MissingModifiedAllan => "missing_modified_allan",
    }
    .to_string()
}

fn power_law_options(
    (
        min_points_per_octave,
        slope_tolerance,
        scatter_tolerance,
        basic_tau_s,
        measurement_bandwidth_hz,
    ): PowerLawOptionsTerm,
) -> PowerLawNoiseOptions {
    let mut options = PowerLawNoiseOptions::new(basic_tau_s, measurement_bandwidth_hz);
    options.min_points_per_octave = min_points_per_octave;
    options.slope_tolerance = slope_tolerance;
    options.scatter_tolerance = scatter_tolerance;
    options
}

fn power_law_error_atom(error: PowerLawNoiseError) -> rustler::Atom {
    match error {
        PowerLawNoiseError::InvalidOptions { .. } => atoms::invalid_options(),
        PowerLawNoiseError::InvalidCurve { .. } => atoms::invalid_curve(),
    }
}

fn octave_term(value: PowerLawOctave) -> PowerLawOctaveTerm {
    let (dominance, noise_type, flag) = match value.dominance {
        PowerLawOctaveDominance::Dominant(noise_type) => (
            "dominant".to_string(),
            Some(noise_type_string(noise_type)),
            None,
        ),
        PowerLawOctaveDominance::Ambiguous => ("ambiguous".to_string(), None, None),
        PowerLawOctaveDominance::Flagged(flag) => {
            ("flagged".to_string(), None, Some(flag_string(flag)))
        }
    };
    PowerLawOctaveTerm {
        tau_start_s: value.tau_start_s,
        tau_end_s: value.tau_end_s,
        point_count: value.point_count as i64,
        adev_slope: value.adev_slope,
        mdev_slope: value.mdev_slope,
        slope_scatter: value.slope_scatter,
        dominance,
        noise_type,
        flag,
    }
}

fn region_term(value: PowerLawNoiseRegion) -> PowerLawRegionTerm {
    PowerLawRegionTerm {
        noise_type: noise_type_string(value.noise_type),
        tau_start_s: value.tau_start_s,
        tau_end_s: value.tau_end_s,
        octave_count: value.octave_count as i64,
        point_count: value.point_count as i64,
        mean_slope: value.mean_slope,
        coefficient: value.coefficient,
    }
}

fn fit_term(value: PowerLawNoiseFit) -> PowerLawFitTerm {
    PowerLawFitTerm {
        dominant_per_octave: value
            .dominant_per_octave
            .into_iter()
            .map(octave_term)
            .collect(),
        coefficients: value.coefficients.into_iter().collect(),
        regions: value.regions.into_iter().map(region_term).collect(),
    }
}

fn compute_one(
    estimator_kind: &str,
    series: AllanSeries<'_>,
    tau0_s: f64,
    averaging_factors: &[usize],
) -> Result<AllanResult, AllanError> {
    match estimator(estimator_kind).map_err(|_| AllanError::NoEstimators)? {
        AllanEstimator::Adev => allan_deviation(series, tau0_s, averaging_factors),
        AllanEstimator::OverlappingAdev => overlapping_adev(series, tau0_s, averaging_factors),
        AllanEstimator::Mdev => modified_adev(series, tau0_s, averaging_factors),
        AllanEstimator::Hdev => hadamard_deviation(series, tau0_s, averaging_factors),
        AllanEstimator::Tdev => time_deviation(series, tau0_s, averaging_factors),
    }
}

/// Compute one explicit Allan-family estimator curve.
#[rustler::nif(schedule = "DirtyCpu")]
fn clock_allan_estimator<'a>(
    env: Env<'a>,
    estimator_kind: String,
    series_kind: String,
    samples: Vec<Option<f64>>,
    tau0_s: f64,
    averaging_factors: Vec<usize>,
) -> NifResult<Term<'a>> {
    let storage = series_owned(&series_kind, samples)?;
    let series = storage.as_series(&series_kind)?;
    Ok(
        match compute_one(&estimator_kind, series, tau0_s, &averaging_factors) {
            Ok(result) => (atoms::ok(), result_term(result)).encode(env),
            Err(error) => (atoms::error(), error_atom(error)).encode(env),
        },
    )
}

/// Compute a configured set of Allan-family curves.
#[rustler::nif(schedule = "DirtyCpu")]
#[allow(clippy::too_many_arguments)]
fn clock_compute_allan_deviations<'a>(
    env: Env<'a>,
    series_kind: String,
    samples: Vec<Option<f64>>,
    tau0_s: f64,
    estimators: (bool, bool, bool, bool, bool),
    tau_grid_kind: String,
    tau_grid_factors: Vec<usize>,
    gap_policy_kind: String,
) -> NifResult<Term<'a>> {
    let storage = series_owned(&series_kind, samples)?;
    let series = storage.as_series(&series_kind)?;
    let mut allan_options = AllanOptions::default();
    allan_options.estimators = estimator_set(estimators);
    allan_options.tau_grid = tau_grid(&tau_grid_kind, tau_grid_factors)?;
    allan_options.gap_policy = gap_policy(&gap_policy_kind)?;
    let input = AllanInput {
        series,
        tau0_s,
        options: allan_options,
    };
    Ok(match compute_allan_deviations(&input) {
        Ok(curves) => (atoms::ok(), curves_term(curves)).encode(env),
        Err(error) => (atoms::error(), error_atom(error)).encode(env),
    })
}

/// Extract RINEX receiver-clock offsets as phase deviations in seconds.
#[rustler::nif(schedule = "DirtyCpu")]
fn clock_receiver_phase_deviations(handle: ResourceArc<RinexObsResource>) -> Vec<Option<f64>> {
    receiver_clock_phase_deviations(&handle.obs)
}

/// Return the exact log-log slope for a power-law noise type.
#[rustler::nif]
fn clock_power_law_slope(noise_type_kind: String, estimator_kind: String) -> NifResult<f64> {
    let noise_type = noise_type(&noise_type_kind)?;
    match estimator_kind.as_str() {
        "adev" | "overlapping_adev" => Ok(allan_deviation_power_law_slope(noise_type)),
        "mdev" => Ok(modified_allan_deviation_power_law_slope(noise_type)),
        _ => Err(Error::Term(Box::new("unknown power-law slope estimator"))),
    }
}

/// Identify power-law clock noise from supplied ADEV and MDEV curves.
#[rustler::nif(schedule = "DirtyCpu")]
fn clock_fit_power_law_noise<'a>(
    env: Env<'a>,
    adev: AllanResultTerm,
    mdev: AllanResultTerm,
    opts: PowerLawOptionsTerm,
) -> NifResult<Term<'a>> {
    let adev = result_from_term(adev)?;
    let mdev = result_from_term(mdev)?;
    Ok(
        match fit_power_law_noise(&adev, &mdev, power_law_options(opts)) {
            Ok(fit) => (atoms::ok(), fit_term(fit)).encode(env),
            Err(error) => (atoms::error(), power_law_error_atom(error)).encode(env),
        },
    )
}
